File: src/aump_conformance/reports.py
"""Report rendering for AUMP conformance runs."""

from __future__ import annotations

from xml.etree import ElementTree

from aump_conformance.jsonio import dump_json
from aump_conformance.models import SuiteReport


def render_text(report: SuiteReport) -> str:
    """Render a compact terminal report."""
    lines = [
        f"{report.suite} v{report.version} (spec {report.spec_version})",
        f"{report.passed}/{report.total} passed",
    ]
    for result in report.results:
        status = "PASS" if result.passed else "FAIL"
        lines.append(f"{status} {result.id} - {result.title}")
        if not result.passed and result.message:
            lines.append(f"  {result.message}")
        if not result.passed:
            lines.append(f"  expected: {result.expected}")
            lines.append(f"  actual: {result.actual}")
    return "\n".join(lines) + "\n"


def render_json(report: SuiteReport) -> str:
    """Render a JSON report."""
    return dump_json(report.to_dict())


def render_junit(report: SuiteReport) -> str:
    """Render a JUnit XML report."""
    suite = ElementTree.Element(
        "testsuite",
        {
            "name": report.suite,
            "tests": str(report.total),
            "failures": str(report.failed),
        },
    )
    for result in report.results:
        case = ElementTree.SubElement(
            suite,
            "testcase",
            {
                "classname": result.category,
                "name": result.id,
            },
        )
        if not result.passed:
            failure = ElementTree.SubElement(
                case,
                "failure",
                {
                    "message": result.message or "conformance case failed",
                },
            )
            failure.text = dump_json(result.to_dict())

    return ElementTree.tostring(suite, encoding="unicode") + "\n"

File: internal/conformance/reports.go
package conformance

import (
	"encoding/xml"
	"fmt"
	"strings"
)

func RenderText(report SuiteReport) string {
	var builder strings.Builder
	builder.WriteString(fmt.Sprintf("%s v%s (spec %s)\n", report.Suite, report.Version, report.SpecVersion))
	builder.WriteString(fmt.Sprintf("%d/%d passed\n", report.Passed(), report.Total()))
	for _, result := range report.Results {
		status := "PASS"
		if !result.Passed {
			status = "FAIL"
		}
		builder.WriteString(fmt.Sprintf("%s %s - %s\n", status, result.ID, result.Title))
		if !result.Passed && result.Message != "" {
			builder.WriteString("  " + result.Message + "\n")
		}
		if !result.Passed {
			builder.WriteString(fmt.Sprintf("  expected: %v\n", result.Expected))
			builder.WriteString(fmt.Sprintf("  actual: %v\n", result.Actual))
		}
	}
	return builder.String()
}

func RenderJSON(report SuiteReport) (string, error) {
	return dumpJSON(report.JSON())
}

func RenderJUnit(report SuiteReport) (string, error) {
	type failure struct {
		Message string `xml:"message,attr"`
		Text    string `xml:",chardata"`
	}
	type testcase struct {
		Classname string   `xml:"classname,attr"`
		Name      string   `xml:"name,attr"`
		Failure   *failure `xml:"failure,omitempty"`
	}
	type testsuite struct {
		XMLName  xml.Name   `xml:"testsuite"`
		Name     string     `xml:"name,attr"`
		Tests    int        `xml:"tests,attr"`
		Failures int        `xml:"failures,attr"`
		Cases    []testcase `xml:"testcase"`
	}

	suite := testsuite{Name: report.Suite, Tests: report.Total(), Failures: report.Failed()}
	for _, result := range report.Results {
		tc := testcase{Classname: result.Category, Name: result.ID}
		if !result.Passed {
			payload, err := dumpJSON(result)
			if err != nil {
				return "", err
			}
			message := result.Message
			if message == "" {
				message = "conformance case failed"
			}
			tc.Failure = &failure{Message: message, Text: payload}
		}
		suite.Cases = append(suite.Cases, tc)
	}

	data, err := xml.Marshal(suite)
	if err != nil {
		return "", err
	}
	return string(data) + "\n", nil
}
